File: experiments/prioritized_search.py
"""Planning-only controlled best-first allocation experiment."""
import argparse,copy,hashlib,json,sys,time
from collections import defaultdict
from pathlib import Path
ROOT=Path(__file__).resolve().parents[1];sys.path.insert(0,str(ROOT))
from graph_runtime import execute_graph,ExecutionCache
from experiments.adaptive_reuse import graph
from experiments.compositional_transfer import prepare as previous_prepare
from experiments.transfer_tasks import VOCABULARIES
ARMS=['no_memory','managed_memory']
SEEDS=[7109,8111]
BUDGET=450000

def prepare(database,arm):
 store,lib,hub,cost=previous_prepare(database,'compositional')
 lib.update(json.loads((ROOT/'curriculum/prioritized-search.json').read_text()))
 return store,lib,hub,cost

def prioritized_search(lib,hub,request,budget,vocabulary):
 cache=ExecutionCache();costs=defaultdict(int);log=[];diagnostics=[]
 cpu=time.process_time();wall=time.perf_counter();status='step_budget'
 def step(name,arg,category):
  meter={}
  try:return execute_graph(lib[name]['graph'],arg,lib,limit=budget-sum(costs.values()),sensors=hub,execution_cache=cache,meter=meter)[0]
  except ValueError:
   if not meter.get('budget_exhausted'):raise
   return None
  finally:costs[category]+=meter.get('logical_steps',0)
 state=step('priority_init',{**request,'base':vocabulary},'initialization')
 if state is None:state={'done':False,'found':None,'selected':[],'evaluated':0}
 else:
  while not state['done'] and sum(costs.values())<budget:
   if time.process_time()-cpu>=10:status='cpu_budget';break
   result=step('priority_step',state,'allocation_and_execution')
   if result is None:break
   state=result;log+=result['last_log']
   diagnostics.append({'ops':result['last'],'expanded':result['last_expanded'],'allowed':result['last_allowed'],'duplicate':result['last_duplicate'],'progress':result['last_evaluation']['progress'],'matches':result['last_evaluation']['matches'],'executable':result['last_evaluation']['executable'],'queue_size':len(result['queue'])})
 if state['found'] is not None:status='found'
 elif state['done']:status='exhausted_space'
 return {'status':status,'found':state['found'],'graph_steps':sum(costs.values()),'costs':dict(costs),'log':log,'selected':state['selected'],'candidates_evaluated':state['evaluated'],'mixed_candidates_evaluated':sum(x['allowed'] and len(x['ops'])>1 and any(op not in vocabulary for op in x['ops']) for x in diagnostics),'max_call_depth':max((len(x['ops']) for x in diagnostics if x['allowed']),default=0),'diagnostics':diagnostics,'cpu_seconds':time.process_time()-cpu,'wall_seconds':time.perf_counter()-wall}

def trial(store,lib,hub,task,arm):
 cpu=time.process_time();wall=time.perf_counter();changes=store.connection.total_changes;costs=defaultdict(int)
 request=copy.deepcopy(task['request']);remaining=BUDGET;attempts=[];logs=[];found=None;validation={'accepted':False};rejections=0
 vocabulary=VOCABULARIES[task['domain']];runner=lambda lib,hub,request,budget: prioritized_search(lib,hub,request,budget,vocabulary)
 for attempt in range(2):
  if remaining<=0:break
  result=runner(lib,hub,request,remaining)
  attempts.append(result);logs+=result['log'];found=result['found'];remaining-=result['graph_steps']
  for k,v in result['costs'].items():costs[k]+=v
  validation,cost=graph(lib,hub,'economy_validate',{'ops':found,'examples':task['validation'],'predicate':'unused','projector':'unused'});costs['validation']+=cost['graph_steps']
  if found is None or validation['accepted']:break
  rejections+=1
  if attempt==0 and remaining>0:
   request['examples'],cost=graph(lib,hub,'economy_refine',{'examples':request['examples'],'validation':task['validation']});costs['feedback']+=cost['graph_steps'];remaining-=cost['graph_steps']
 retained={'retained':False}
 if arm in ['managed_memory']:
  _,cost=graph(lib,hub,'manager_update',{'log':logs,'found':found,'validated':validation['accepted']});costs['utility']+=cost['graph_steps']
  if validation['accepted']:
   retained,cost=graph(lib,hub,'manager_retain',{'ops':found,'name':'allocated_'+str(task['id']),'base':vocabulary,'validation':task['validation'],'predicate':'unused','projector':'unused'});costs['retention']+=cost['graph_steps']
 audit={'accepted':False}
 if validation['accepted']:
  audit,cost=graph(lib,hub,'economy_validate',{'ops':found,'examples':task['audit'],'predicate':'unused','projector':'unused'});costs['audit']+=cost['graph_steps']
 catalog=store.map('knowledge.economics')['catalog']
 if arm in ['no_memory']:assert not catalog
 dbbytes=store.connection.execute('PRAGMA page_count').fetchone()[0]*store.connection.execute('PRAGMA page_size').fetchone()[0]
 return {'task':task['id'],'domain':task['domain'],'family':task['family'],'expected_boundary':task['expected_boundary'],'arm':arm,'found':found,'success':bool(validation['accepted'] and audit['accepted']),'composed_solution':bool(found and len(found)>1 and any(op not in vocabulary for op in found)),'validation_rejections':rejections,'audit_false_positive':bool(validation['accepted'] and not audit['accepted']),'retained':retained['retained'],'catalog':catalog,'library_size':len(catalog),'attempts':attempts,'costs':dict(costs),'total_graph_steps':sum(costs.values()),'cpu_seconds':time.process_time()-cpu,'wall_seconds':time.perf_counter()-wall,'database_bytes':dbbytes,'sqlite_row_changes':store.connection.total_changes-changes}

def freeze(directory):
 directory=Path(directory);directory.mkdir(parents=True,exist_ok=True)
 if (directory/'manifest.json').exists():raise ValueError('Frozen output already exists.')
 parent=ROOT/'experiments/compositional-transfer-results/manifest.json'
 old=json.loads(parent.read_text())
 # Preserve every previously frozen dependency, including all memory rules.
 for p,h in old['source_hashes'].items():assert hashlib.sha256((ROOT/p).read_bytes()).hexdigest()==h,p
 paths=list(old['source_hashes'])+['experiments/prioritized_search.py','curriculum/prioritized-search.json','graph-authoring/prioritized-search.mjs']
 manifest={'seeds':SEEDS,'arms':ARMS,'budget':BUDGET,'max_depth':5,'max_attempts':2,'streams':{'planning':old['streams']['planning']},'parent_manifest_sha256':hashlib.sha256(parent.read_bytes()).hexdigest(),'source_hashes':{p:hashlib.sha256((ROOT/p).read_bytes()).hexdigest() for p in paths},'policy':'Best-first queue capped at 100 partial programs. Complete all children of the selected parent, then choose lowest priority. Integer priority = expanded primitive length - 1000000 * full training-example matches - floor(1000 * sum of per-example fractions of changed target scalar leaves already matched). Scalar target leaves equal to input at the same path are excluded from heuristic only. Full output equality remains acceptance. Exact expanded-sequence dedup; failed executable prefixes pruned; successful prefixes below expanded length five queued. All callables use the same expansion rule. Up to four recent active methods selected by unchanged memory manager. No whole-program probe phase.','limitations':['Planning only, two previously evaluated streams: architectural diagnostic, not fresh confirmation or cross-domain transfer evidence.','Heuristic is newly supplied generic search machinery. No planning action or field names occur in allocator rules.','Partial scoring ignores unchanged target leaves and empty containers; full equality still validates solutions.','Priority is heuristic, not an admissible remaining-cost estimate. Beam truncation may discard useful candidates.','All scoring, canonicalization, queue and execution graph work is charged; host CPU, wall time and storage are separately reported.','Memory canonicalization, retention, routing and utility rules unchanged. No seeded learned methods.','No post-freeze tuning; text/tree untouched.']}
 data=json.dumps(manifest,sort_keys=True).encode();(directory/'manifest.json').write_bytes(data);(directory/'manifest.sha256').write_text(hashlib.sha256(data).hexdigest()+'\n')

def run(directory):
 directory=Path(directory);manifest=json.loads((directory/'manifest.json').read_text());digest=hashlib.sha256((directory/'manifest.json').read_bytes()).hexdigest()
 if (directory/'trials.jsonl').exists():raise ValueError('Refusing to overwrite an existing run.')
 rows=[];setups=[]
 with (directory/'trials.jsonl').open('w') as out:
  for domain in manifest['streams']:
   for seed in SEEDS:
    engines={}
    try:
     for arm in ARMS:
      cpu=time.process_time();wall=time.perf_counter();store,lib,hub,cost=prepare(directory/f'{domain}-{seed}-{arm}.sqlite3',arm);engines[arm]=(store,lib,hub)
      setups.append({'domain':domain,'seed':seed,'arm':arm,'graph_steps':cost['graph_steps'],'cpu_seconds':time.process_time()-cpu,'wall_seconds':time.perf_counter()-wall,'database_bytes':store.connection.execute('PRAGMA page_count').fetchone()[0]*store.connection.execute('PRAGMA page_size').fetchone()[0]})
     for task in manifest['streams'][domain][str(seed)]:
      offset=(task['id']+seed)%len(ARMS);order=ARMS[offset:]+ARMS[:offset]
      for arm in order:
       store,lib,hub=engines[arm];row=trial(store,lib,hub,task,arm)|{'seed':seed};rows.append(row);out.write(json.dumps(row)+'\n');out.flush()
      if (task['id']+1)%10==0:print(domain,seed,task['id']+1,'/30',{a:sum(r['success'] for r in rows if r['domain']==domain and r['seed']==seed and r['arm']==a) for a in ARMS},flush=True)
    finally:
     for store,_,_ in engines.values():store.close()
 assert hashlib.sha256((directory/'manifest.json').read_bytes()).hexdigest()==digest
 for p,h in manifest['source_hashes'].items():assert hashlib.sha256((ROOT/p).read_bytes()).hexdigest()==h,p
 (directory/'report.json').write_text(json.dumps({'manifest_sha256':digest,'trials':rows,'setup':setups},indent=2)+'\n')

if __name__=='__main__':
 p=argparse.ArgumentParser();p.add_argument('--output',required=True);p.add_argument('--freeze',action='store_true');a=p.parse_args()
 if a.freeze:freeze(a.output);print('Planning allocator experiment frozen.')
 else:run(a.output)

File: graph-authoring/prioritized-search.mjs
// Generic best-first allocation over Data -> Data callables. No domain names.
import fs from 'node:fs';import {G} from './graph.mjs';
const suite={};const save=(name,g,x)=>suite[name]={graph:{...g.finish(x),execution_budget:10000000},source:'Generic changed-target-leaf best-first search; supplied allocator, not learned domain knowledge.'};
const call=(g,name,x)=>g.op('call',[x],{name});
// Sum a list of Data numbers, entirely through graph operations.
let guard=new G(),body=new G(),g=new G();
let more=guard.lt(guard.get(guard.input,'i'),guard.len(guard.get(guard.input,'values')));
let next=body.rec({values:body.get(body.input,'values'),i:body.calc('add',body.get(body.input,'i'),body.data(1)),total:body.calc('add',body.get(body.input,'total'),body.item(body.get(body.input,'values'),body.get(body.input,'i')))});
let loop=g.op('while',[g.rec({values:g.input,i:g.data(0),total:g.data(0)})],{guard:guard.finish(more,'Bool'),body:body.finish(next)});save('priority_sum',g,g.get(loop,'total'));
// Flatten any target tree to scalar paths. No selected field names or types.
const child=new G(),cx=child.get(child.input,'context'),key=child.get(child.input,'item');
const childValue=child.rec({value:child.item(child.get(cx,'value'),key),path:child.push(child.get(cx,'path'),key)});
guard=new G();body=new G();g=new G();
more=guard.lt(guard.data(0),guard.len(guard.get(guard.input,'todo')));
const entry=body.item(body.get(body.input,'todo'),body.data(0)),value=body.get(entry,'value'),kind=body.op('kind_of',[value]);
const compound=body.op('or',[body.eq(kind,body.data('record')),body.eq(kind,body.data('list'))],{},'Bool');
const keys=body.choose(body.eq(kind,body.data('record')),body.op('keys',[value]),body.op('indices',[value]));
const remaining=body.op('slice',[body.get(body.input,'todo')],{start:1});
next=body.rec({todo:body.choose(compound,body.concat(body.op('map',[keys,entry],{body:child.finish(childValue)}),remaining),remaining),leaves:body.choose(compound,body.get(body.input,'leaves'),body.push(body.get(body.input,'leaves'),entry))});
loop=g.op('while',[g.rec({todo:g.list(g.input),leaves:g.data([])})],{guard:guard.finish(more,'Bool'),body:body.finish(next)});save('priority_leaves',g,g.get(loop,'leaves'));
// Safe arbitrary path lookup, including missing paths in intermediate outputs.
guard=new G();body=new G();g=new G();
more=guard.lt(guard.get(guard.input,'i'),guard.len(guard.get(guard.input,'path')));
next=body.rec({path:body.get(body.input,'path'),i:body.calc('add',body.get(body.input,'i'),body.data(1)),value:body.item(body.get(body.input,'value'),body.item(body.get(body.input,'path'),body.get(body.input,'i')))});
loop=g.op('while',[g.rec({path:g.get(g.input,'path'),value:g.get(g.input,'value'),i:g.data(0)})],{guard:guard.finish(more,'Bool'),body:body.finish(next)});save('priority_lookup',g,g.get(loop,'value'));
const attemptBody=new G();const looked=call(attemptBody,'priority_lookup',attemptBody.input);
const feature=new G(),f=feature.get(feature.input,'item');
let attempt=feature.op('attempt',[feature.rec({value:feature.get(feature.input,'context'),path:feature.get(f,'path')})],{body:attemptBody.finish(looked)});
const unchanged=feature.choose(feature.bool(feature.get(attempt,'ok')),feature.datum(feature.eq(feature.get(attempt,'result'),feature.get(f,'value'))),feature.data(false));
g=new G();save('priority_features',g,g.rec({input:g.get(g.input,'input'),expected:g.get(g.input,'expected'),features:g.op('filter',[call(g,'priority_leaves',g.rec({value:g.get(g.input,'expected'),path:g.data([])})),g.get(g.input,'input')],{body:feature.finish(feature.not(feature.bool(unchanged)),'Bool')})}));
// Score a successful intermediate output against changed target leaves.
const match=new G(),mf=match.get(match.input,'item');
attempt=match.op('attempt',[match.rec({value:match.get(match.input,'context'),path:match.get(mf,'path')})],{body:attemptBody.finish(looked)});
const equal=match.choose(match.bool(match.get(attempt,'ok')),match.datum(match.eq(match.get(attempt,'result'),match.get(mf,'value'))),match.data(false));
const one=match.choose(match.bool(equal),match.data(1),match.data(0));
const scoreCase=new G(),sc=scoreCase.get(scoreCase.input,'context'),ix=scoreCase.get(scoreCase.input,'item');
const observation=scoreCase.item(scoreCase.get(sc,'cases'),ix),features=scoreCase.get(scoreCase.item(scoreCase.get(sc,'examples'),ix),'features');
const score=scoreCase.choose(scoreCase.lt(scoreCase.data(0),scoreCase.len(features)),scoreCase.calc('divide',call(scoreCase,'priority_sum',scoreCase.op('map',[features,scoreCase.get(observation,'execution','result')],{body:match.finish(one)})),scoreCase.len(features)),scoreCase.data(0));
const safeScore=scoreCase.choose(scoreCase.bool(scoreCase.get(observation,'execution','ok')),score,scoreCase.data(0));
const flag=new G(),valid=new G();
g=new G();const evaluation=call(g,'synth_evaluate_candidate',g.input),cases=g.get(evaluation,'cases');
const matches=call(g,'priority_sum',g.op('map',[cases],{body:flag.finish(flag.choose(flag.bool(flag.get(flag.input,'passed')),flag.data(1),flag.data(0)))}));
const progress=call(g,'priority_sum',g.op('map',[g.op('indices',[cases]),g.rec({cases,examples:g.get(g.input,'examples')})],{body:scoreCase.finish(safeScore)}));
const executable=g.not(g.contains(g.op('map',[cases],{body:valid.finish(valid.get(valid.input,'execution','ok'))}),g.data(false)));
save('priority_evaluate',g,g.rec({accepted:g.get(evaluation,'accepted'),executable:g.datum(executable),progress,matches}));
const enrich=new G();g=new G();const prepared=call(g,'composition_prepare',g.rec({base:g.get(g.input,'base'),compose:g.data(true)}));
save('priority_init',g,g.rec({vocabulary:g.get(prepared,'vocabulary'),catalog:g.get(prepared,'catalog'),selected:g.get(prepared,'selected'),examples:g.op('map',[g.get(g.input,'examples')],{body:enrich.finish(call(enrich,'priority_features',enrich.input))}),predicate:g.get(g.input,'predicate'),projector:g.get(g.input,'projector'),max_depth:g.data(5),beam:g.data(100),parent:g.data({ops:[],priority:0}),next_i:g.data(0),queue:g.data([]),seen:g.data([]),found:g.data(null),done:g.data(false),evaluated:g.data(0)}));
// One child evaluation per resumable graph transition. Finish a parent's
// children, then expand the best queued partial program, regardless of depth.
const logged=new G();g=new G();const s=g.input;
const ops=g.push(g.get(s,'parent','ops'),g.item(g.get(s,'vocabulary'),g.get(s,'next_i')));
const canonical=g.get(call(g,'manager_canonicalize',g.rec({ops,catalog:g.get(s,'catalog')})),'ops');
const duplicate=g.contains(g.get(s,'seen'),canonical),length=g.len(canonical);
const allowed=g.and(g.not(duplicate),g.not(g.lt(g.get(s,'max_depth'),length)));
const ev=g.choose(allowed,call(g,'priority_evaluate',g.rec({ops,examples:g.get(s,'examples'),predicate:g.get(s,'predicate'),projector:g.get(s,'projector')})),g.data({accepted:false,executable:false,progress:0,matches:0}));
// Exact-example matches dominate; target-leaf progress follows. Lower expanded
// cost breaks ties, then stable generation order. No action-specific heuristics.
const scaled=g.op('as_data',[g.op('floor',[g.num(g.calc('multiply',g.get(ev,'progress'),g.data(1000)))],{},'Number')]);
const priority=g.calc('subtract',length,g.calc('add',g.calc('multiply',g.get(ev,'matches'),g.data(1000000)),scaled));
const candidate=g.rec({ops,priority});
const queued=g.choose(g.and(g.bool(g.get(ev,'executable')),g.lt(length,g.get(s,'max_depth'))),g.push(g.get(s,'queue'),candidate),g.get(s,'queue'));
const sorted=g.op('slice',[g.op('sort',[queued],{key:'priority'})],{start:0,stop:100});
const end=g.eq(g.calc('add',g.get(s,'next_i'),g.data(1)),g.len(g.get(s,'vocabulary'))),empty=g.eq(g.len(sorted),g.data(0));
const done=g.op('or',[g.bool(g.get(ev,'accepted')),g.and(end,empty)],{},'Bool');
save('priority_step',g,g.rec({vocabulary:g.get(s,'vocabulary'),catalog:g.get(s,'catalog'),selected:g.get(s,'selected'),examples:g.get(s,'examples'),predicate:g.get(s,'predicate'),projector:g.get(s,'projector'),max_depth:g.get(s,'max_depth'),beam:g.get(s,'beam'),parent:g.choose(g.and(end,g.not(empty)),g.item(sorted,g.data(0)),g.get(s,'parent')),next_i:g.choose(end,g.data(0),g.calc('add',g.get(s,'next_i'),g.data(1))),queue:g.choose(g.and(end,g.not(empty)),g.op('slice',[sorted],{start:1}),sorted),seen:g.choose(allowed,g.push(g.get(s,'seen'),canonical),g.get(s,'seen')),found:g.choose(g.bool(g.get(ev,'accepted')),ops,g.data(null)),done:g.datum(done),evaluated:g.calc('add',g.get(s,'evaluated'),g.choose(allowed,g.data(1),g.data(0))),last:ops,last_expanded:canonical,last_duplicate:g.datum(duplicate),last_allowed:g.datum(allowed),last_evaluation:ev,last_log:g.choose(allowed,g.op('map',[ops],{body:logged.finish(logged.rec({ops:logged.list(logged.input)}))}),g.data([]))}));
fs.writeFileSync(new URL('../curriculum/prioritized-search.json',import.meta.url),JSON.stringify(suite,null,2)+'\n');
